include/deepstream.hpp

```diff
@@ -68,32 +90,156 @@ namespace deepstream {
+         * @param[in] callback A LoginCallback that will be invoked with the
+         * user data on successful login.
          * in `p_user_data` if the reference is not `NULL`.
          */
-        void login(const std::string &auth, const Client::LoginCallback &callback)
+        void login(const json::object_t &auth, const LoginCallback &callback)
         {
-            client_.login(auth, callback);
+            const std::string auth_str(json(auth).dump());
+            const Buffer auth_buff(auth_str);
+
+            const Client::LoginCallback core_callback([callback, this](const Buffer &&login_data) {
+                const json data = json_handler_.prefixed_to_json(login_data);
+                callback(std::move(data));
+            });
+
+            client_.login(auth_buff, core_callback);
         }
 
         void close();
 
-        ConnectionState get_connection_state(){
+        ConnectionState get_connection_state()
+        {
             return client_.get_connection_state();
         }
 
-
     private:
-
         PocoWSHandler wsh_;
         BasicErrorHandler error_handler_;
         Client client_;
 
+        struct EventWrapper {
+            typedef std::function<void(const json &)> SubscribeFn;
+            typedef std::function<bool(const std::string &, bool)> ListenFn;
+
+            EventWrapper() = delete;
+
+            EventWrapper(const EventWrapper &) = delete;
+
+            EventWrapper &operator=(const EventWrapper &) = delete;
+
+            EventWrapper(Client &client, ErrorHandler &error_handler, JSONHandler &json_handler)
+                : client_(client)
+                , error_handler_(error_handler)
+                , json_handler_(json_handler)
+            {
+            }
+
+            void emit(std::string name, json data)
+            {
+                const Buffer &data_buff = json_handler_.to_prefixed_buffer(data);
+                const Buffer name_buff(name);
+                client_.event.emit(name_buff, data_buff);
+            }
+
+            SubscriptionId subscribe(std::string name, SubscribeFn callback)
+            {
+                Buffer name_buff(name);
+                Event::SubscribeFn core_callback([callback, this](const Buffer &prefixed_buff) {
+                    const json &data = json_handler_.prefixed_to_json(prefixed_buff);
+                    callback(data);
+                });
+                SubscriptionId subscription_id = client_.event.subscribe(name_buff, core_callback);
+                return subscription_id;
+            }
+
+            void unsubscribe(std::string name, SubscriptionId subscription_id)
+            {
+                Buffer name_buff(name);
+                client_.event.unsubscribe(name_buff, subscription_id);
+            }
+
+            void listen(const std::string &pattern, const ListenFn callback)
+            {
+                Event::ListenFn core_callback([callback](const Event::Name &match, bool is_subscribed) -> bool {
```

Comment:
We map from core's `Buffer` type (which is a lightweight wrapper around vector<char>) to std::string. Having to do this isn't ideal and *might* hurt performance to some extent. I'm considering the possibility that we add implicit `Buffer(std::string s)` and `std::string(Buffer buff)` constructors, so that conversions happen automatically. The argument against that is that it makes it harder to tell when a copy is going to happen (since it is less explicit), and therefore what the performance implications will be.